File: .claude/skills/engineering-insights/scripts/append_insight.py
#!/usr/bin/env python3
"""Append-only writer for <package>/INSIGHTS.md.

  append_insight.py add <package> "<Section>" "<entry text>"   # adds one dated bullet
  append_insight.py dupes <package> "<entry text>"              # lists similar existing bullets
  append_insight.py verify <package>                            # no line removed vs git HEAD

Guarantees for `add`: the file is only ever grown — the new bullet is inserted as the
first bullet of <Section>; every pre-existing line is checked to still be present, in
order, before the file is written. Refuses near-duplicates (exit 3) and unknown sections.
"""
import datetime
import re
import subprocess
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parents[4]
PACKAGES = ("server", "client", "reviewer-core", "e2e")
SECTIONS = (
    "What Works",
    "What Doesn't Work",
    "Codebase Patterns",
    "Tool & Library Notes",
    "Recurring Errors & Fixes",
    "Session Notes",
    "Open Questions",
)
# Share of the new entry's words already present in one existing bullet.
DUPLICATE_THRESHOLD = 0.6
STOPWORDS = set("a an the to of in on for and or is are be it this that with as at by from not use".split())


def fail(code, msg):
    print(f"ERROR: {msg}", file=sys.stderr)
    sys.exit(code)


def insights_path(package):
    if package not in PACKAGES:
        fail(2, f"unknown package '{package}', expected one of {', '.join(PACKAGES)}")
    path = ROOT / package / "INSIGHTS.md"
    if not path.is_file():
        fail(2, f"{path} not found")
    return path


def words(text):
    text = re.sub(r"^\s*-\s*\d{4}-\d{2}-\d{2}\s*[—-]\s*", "", text)
    return {w for w in re.findall(r"[a-z0-9_./-]{2,}", text.lower()) if w not in STOPWORDS}


def similar_bullets(lines, entry):
    new = words(entry)
    if not new:
        return []
    hits = []
    for line in lines:
        if line.lstrip().startswith("- "):
            overlap = len(new & words(line)) / len(new)
            if overlap >= DUPLICATE_THRESHOLD:
                hits.append((overlap, line.strip()))
    return sorted(hits, reverse=True)


def is_subsequence(old, new):
    it = iter(new)
    return all(any(line == candidate for candidate in it) for line in old)


def cmd_add(package, section, entry):
    if section not in SECTIONS:
        fail(2, f"unknown section '{section}', expected one of: {'; '.join(SECTIONS)}")
    entry = " ".join(entry.split())
    if not entry:
        fail(2, "empty entry")
    path = insights_path(package)
    old = path.read_text(encoding="utf-8").splitlines()

    dupes = similar_bullets(old, entry)
    if dupes:
        print("DUPLICATE? Not written. Similar existing entries:")
        for score, line in dupes:
            print(f"  {score:.0%}  {line}")
        sys.exit(3)

    try:
        header = old.index(f"## {section}")
    except ValueError:
        fail(2, f"section '## {section}' missing in {path} — restore the template, don't improvise")
    insert_at = header + 1
    while insert_at < len(old) and old[insert_at].lstrip().startswith("<!--"):
        insert_at += 1

    bullet = f"- {datetime.date.today().isoformat()} — {entry}"
    new = old[:insert_at] + [bullet] + old[insert_at:]
    if not is_subsequence(old, new) or len(new) != len(old) + 1:
        fail(4, "safety check failed: an existing line would change — nothing written")
    path.write_text("\n".join(new) + "\n", encoding="utf-8")
    print(f"ADDED to {package}/INSIGHTS.md › {section}:\n{bullet}")


def cmd_dupes(package, entry):
    hits = similar_bullets(insights_path(package).read_text(encoding="utf-8").splitlines(), entry)
    for score, line in hits:
        print(f"{score:.0%}  {line}")
    if not hits:
        print("no similar entries")


def cmd_verify(package):
    path = insights_path(package)
    rel = path.relative_to(ROOT).as_posix()
    try:
        head = subprocess.run(["git", "-C", str(ROOT), "show", f"HEAD:{rel}"],
                              capture_output=True, text=True, check=True).stdout.splitlines()
    except subprocess.CalledProcessError:
        print(f"OK: {rel} not in HEAD yet — nothing to compare")
        return
    now = path.read_text(encoding="utf-8").splitlines()
    missing = [line for line in head if line not in now]
    if missing or not is_subsequence(head, now):
        print(f"VIOLATION: {rel} lost or reordered committed lines:")
        for line in missing:
            print(f"  - {line}")
        print(f"Restore with: git checkout HEAD -- {rel}  (then re-add only the new bullets)")
        sys.exit(5)
    print(f"OK: {rel} only grew (+{len(now) - len(head)} lines vs HEAD)")


if __name__ == "__main__":
    args = sys.argv[1:]
    if len(args) == 4 and args[0] == "add":
        cmd_add(*args[1:])
    elif len(args) == 3 and args[0] == "dupes":
        cmd_dupes(*args[1:])
    elif len(args) == 2 and args[0] == "verify":
        cmd_verify(args[1])
    else:
        print(__doc__)
        sys.exit(2)

File: .claude/skills/engineering-insights/scripts/append-insight.mjs
#!/usr/bin/env node
// Append-only writer for INSIGHTS.md files — the only way the
// engineering-insights skill writes. It inserts lines and never changes or
// removes an existing one; anything else is refused with exit 1 and the file
// left byte-for-byte as it was.
//
//   node append-insight.mjs <file> "<Section>" [--under "<text>"] <<'EOF'
//   - **YYYY-MM-DD** — claim. Evidence: `path:line`
//   EOF
//
// Every entry outside "Session notes" must cite a `path:line`, or it is refused.
// <file> is relative to the repo root, wherever this runs from. Without
// --under the entry goes to the end of the section; with --under it becomes an
// indented sub-bullet of the one entry in that section containing <text>.

import { lstatSync, readFileSync, renameSync, writeFileSync } from 'node:fs';
import { basename, dirname, isAbsolute, relative, resolve, sep } from 'node:path';
import { fileURLToPath } from 'node:url';

const ROOT = resolve(dirname(fileURLToPath(import.meta.url)), '../../../..');
const USAGE =
  'usage: node append-insight.mjs <file> "<Section>" [--under "<text>"] <<\'EOF\' … EOF';
const ENTRY_START = /^- \*\*\d{4}-\d{2}-\d{2}\*\* — \S/;
const DATED_BULLET = /^(\s*)- \*\*\d{4}-\d{2}-\d{2}\*\* — (.*)$/;
const SECTION_END = /^#{1,2} /;

function fail(message, suffix = ' — nothing written.') {
  process.stderr.write(`append-insight: ${message}${suffix}\n`);
  process.exit(1);
}

const squash = (s) => s.replace(/\s+/g, ' ').trim();

// ── arguments ────────────────────────────────────────────────────────────────
const positional = [];
let under = null;
const argv = process.argv.slice(2);
for (let i = 0; i < argv.length; i++) {
  if (argv[i] === '--under') {
    under = argv[++i];
    if (!under || !under.trim()) fail('--under needs the text of an existing entry');
  } else {
    positional.push(argv[i]);
  }
}
if (positional.length !== 2) fail(USAGE);
const [fileArg, section] = positional;

// ── target file ──────────────────────────────────────────────────────────────
const file = resolve(ROOT, fileArg);
const rel = relative(ROOT, file).split(sep).join('/');
if (rel.startsWith('../') || rel === '..' || isAbsolute(rel)) fail(`${fileArg} is outside the repo`);
if (basename(file) !== 'INSIGHTS.md') fail(`${rel} is not an INSIGHTS.md file`);
if (rel.startsWith('server/clones/')) fail(`${rel} is inside server/clones/, a cloned copy`);
let stat;
try {
  stat = lstatSync(file);
} catch {
  fail(`${rel} does not exist; this script never creates files — ask the user`);
}
if (!stat.isFile()) fail(`${rel} is not a regular file`);

// ── entry ────────────────────────────────────────────────────────────────────
if (process.stdin.isTTY) fail(`pass the entry on stdin. ${USAGE}`);
const entry = readFileSync(0, 'utf8').replace(/\s+$/, '').split('\n').map((l) => l.replace(/\s+$/, ''));
if (!entry[0]) fail('the entry on stdin is empty');
if (!ENTRY_START.test(entry[0])) fail('the entry must start with "- **YYYY-MM-DD** — "');
for (const line of entry.slice(1)) {
  if (!/^ {2,}\S/.test(line)) {
    fail('continuation lines must be indented by 2+ spaces, with no blank lines in between');
  }
}
const claimOf = (lines) =>
  squash([lines[0].match(DATED_BULLET)[2], ...lines.slice(1)].join(' '));
const claim = claimOf(entry);
// Evidence is a backticked `file.ext:line` (`:12-18`, `:4,22` too). The letter
// extension keeps `localhost:3101` or `127.0.0.1:5433` from passing as a path.
// Only "Session notes" lines are exempt.
const EVIDENCE = /`[^`\s]*\.[A-Za-z][A-Za-z0-9]*:\d+(?:[-,]\d+)*`/;
if (section !== 'Session notes' && !EVIDENCE.test(claim)) {
  fail('the entry needs evidence as a backticked `path:line`, e.g. `src/app.ts:42`');
}

// ── current content ──────────────────────────────────────────────────────────
const original = readFileSync(file, 'utf8');
const lines = (original.endsWith('\n') ? original.slice(0, -1) : original).split('\n');

// Every dated bullet (any depth) with its continuation lines, for the duplicate check.
for (let i = 0; i < lines.length; i++) {
  const m = lines[i].match(DATED_BULLET);
  if (!m) continue;
  const block = [lines[i]];
  for (let j = i + 1; j < lines.length; j++) {
    const l = lines[j];
    if (!l.trim() || /^\s*- /.test(l) || SECTION_END.test(l)) break;
    if (l.length - l.trimStart().length < m[1].length + 2) break;
    block.push(l);
  }
  if (claimOf(block) === claim) fail(`${rel}:${i + 1} already says this`);
}

// ── section ──────────────────────────────────────────────────────────────────
const headings = lines.flatMap((l, i) => (l === `## ${section}` ? [i] : []));
if (headings.length !== 1) {
  const known = lines.filter((l) => l.startsWith('## ')).map((l) => `"${l.slice(3)}"`);
  fail(
    `${rel} has ${headings.length ? 'several' : 'no'} "## ${section}" heading(s); ` +
      `sections: ${known.join(', ')}. This script never creates or renames sections`,
  );
}
const head = headings[0];
let end = lines.length;
for (let i = head + 1; i < lines.length; i++) {
  if (SECTION_END.test(lines[i])) {
    end = i;
    break;
  }
}

// ── insertion point ──────────────────────────────────────────────────────────
let at;
let inserted;
if (under === null) {
  let last = head;
  for (let i = head + 1; i < end; i++) if (lines[i].trim()) last = i;
  at = last + 1;
  inserted = last === head ? ['', ...entry] : [...entry];
} else {
  // A top-level entry block: its bullet line plus everything up to the next
  // blank line, top-level line or heading (continuations and sub-bullets).
  const blocks = [];
  for (let i = head + 1; i < end; i++) {
    if (!lines[i].startsWith('- ')) continue;
    let j = i + 1;
    while (j < end && lines[j].trim() && /^\s/.test(lines[j])) j++;
    blocks.push({ start: i, end: j });
  }
  const needle = squash(under);
  const hits = blocks.filter((b) => squash(lines.slice(b.start, b.end).join(' ')).includes(needle));
  if (hits.length !== 1) {
    fail(`--under text matches ${hits.length} entries in "## ${section}"; it must match exactly one`);
  }
  at = hits[0].end;
  inserted = entry.map((l) => `  ${l}`);
}
if (at < lines.length && SECTION_END.test(lines[at])) inserted.push('');

// ── invariant: old lines all present, unchanged, in order ────────────────────
const next = [...lines.slice(0, at), ...inserted, ...lines.slice(at)];
const intact =
  next.length === lines.length + inserted.length &&
  lines.every((l, i) => next[i < at ? i : i + inserted.length] === l);
if (!intact) fail('internal check failed: an existing line would change');

// ── atomic write ─────────────────────────────────────────────────────────────
const output = `${next.join('\n')}\n`;
if (readFileSync(file, 'utf8') !== original) fail(`${rel} changed while this ran; re-run`);
const tmp = `${file}.tmp-${process.pid}`;
writeFileSync(tmp, output, { mode: stat.mode });
renameSync(tmp, file);
if (readFileSync(file, 'utf8') !== output) {
  fail(`${rel} was written but reads back differently`, ' — check it with git diff.');
}

process.stdout.write(
  `+${inserted.length} lines at ${rel}:${at + 1}; removed 0\n` +
    inserted.map((l) => `+ ${l}`).join('\n') +
    '\n' +
    (original.endsWith('\n')
      ? ''
      : `note: the file had no final newline; one was added after line ${lines.length}, ` +
        'so git diff shows that line as -/+ with its text unchanged\n'),
);
